### PR TITLE
[ddwrt] Discover virtual access point radios

### LINKED ISSUE #21177: [ddwrt] Discovery fails when virtual access point is enabled
## Expected Behavior

Create DD-WRT network Thing using router's hostname.  Device Thing is added to inbox.  After accepting device Thing, radio and client Things are added to inbox and can be accepted.

## Current Behavior

When setting up the DD-WRT add-on, I create a network Thing using my router's hostname.  This allows OpenHAB to discover the device Thing for my router.  I accept the inbox entry.  After running a scan for new DD-WRT Things to add, none are added to the inbox.  The log (/var/log/openhab/openhab.log) shows the following error, repeating every 3 seconds:

```
2026-07-15 18:52:37.433 [WARN ] [ddwrt.internal.DDWRTDiscoveryService] - Error during DD-WRT discovery scan: ID segment '14-91-82-35-d3-ae-wlan1.1' contains invalid characters. Each segment of the ID must match the pattern [\w-]*.
java.lang.IllegalArgumentException: ID segment '14-91-82-35-d3-ae-wlan1.1' contains invalid characters. Each segment of the ID must match the pattern [\w-]*.
        at org.openhab.core.common.AbstractUID.validateSegment(AbstractUID.java:105)
        at org.openhab.core.common.AbstractUID.<init>(AbstractUID.java:76)
        at org.openhab.core.common.AbstractUID.<init>(AbstractUID.java:59)
        at org.openhab.core.thing.UID.<init>(UID.java:56)
        at org.openhab.core.thing.ThingUID.<init>(ThingUID.java:57)
        at org.openhab.binding.ddwrt.internal.DDWRTDiscoveryService.lambda$3(DDWRTDiscoveryService.java:174)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1117)
        at org.openhab.binding.ddwrt.internal.DDWRTDiscoveryService.discoverRadios(DDWRTDiscoveryService.java:166)
        at org.openhab.binding.ddwrt.internal.DDWRTDiscoveryService.startScan(DDWRTDiscoveryService.java:78)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```

## Possible Solution

It appears that in the radio discovery function of DDWRTDiscoveryService.java, the UID is generated from the MAC address and interface name.  Interface names for virtual access points in DD-WRT can have a '.' character, which is forbidden in Thing UIDs.  It is likely that adding a .replace('.', '-') to the get interface name function would allow discovery to run correctly when the DD-WRT configuration includes a virtual AP.

## Steps to Reproduce (for Bugs)

1. Configure DD-WRT router to have at least one virtual access point (E.G. guest network)
2. Create Thing for DD-WRT network, configuring hostname
3. Discovery service creates entry in inbox for device
4. Accept device as new Thing
5. Logs show error during discovery scan

## Context

I am trying to add Things for certain WiFi client devices so I can use them for presence detection.

## Your Environment

* Version used: OpenHAB 5.2.0, DD-WRT binding 5.2.0
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...): DD-WRT build v3.0-r65574 std
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Ubuntu 24.04

### PR DESCRIPTION
**Pull request by ballle98**
## Description

This bugfix enables DD-WRT virtual access point (VAP) radios to be discovered correctly.

Virtual interfaces such as `wlan0.1` and `wl0.1` contain a dot, which is not valid in an openHAB Thing UID segment. Discovery therefore failed while constructing the radio Thing UID. Broadcom VAPs also were not enumerated because the binding only probed a fixed set of physical radio interfaces.

The change:

- sanitizes radio interface names when constructing Thing UIDs while preserving the original dotted interface name in the radio's `interfaceId`;
- reads the Broadcom `wl0_vifs`, `wl1_vifs`, and `wl2_vifs` NVRAM values and safely adds their interfaces to radio discovery;
- expands the physical Broadcom candidates to cover a third radio; and
- adds unit tests for dotted Thing UIDs and Broadcom VAP enumeration.

This allows users with guest networks or other VAPs to discover and monitor those radio interfaces.

Fixes #21177

## Testing

- `mvn clean install` completed successfully for the binding (66 tests).
- Verified on a Marvell DD-WRT access point: `wlan0.1` was discovered as a valid radio Thing with SSID `dd-wrt_vap`.
- Verified on a Broadcom DD-WRT access point: `wl0.1` was discovered with SSID `dd-wrt_vap_m`, and associated-client, RSSI, and noise data were read successfully.
- Confirmed the runtime log no longer contains the invalid Thing UID or discovery-scan errors from this issue.

### CODE REVIEW

File: bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java [hunk 1]
```diff
@@ -116,7 +120,16 @@ protected List<DDWRTRadio> enumerateRadios(SshRunner runner) {
         // Use cached interfaces if available, otherwise probe all candidates
         String[] ifaces = discoveredIfaces;
         if (ifaces == null) {
```
Comment (wborn) [inline bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java:124]: One possible follow-up: after the first successful enumeration, `discoveredIfaces` prevents the `wl*_vifs` values from being read again. A VAP added later, or missed during its first SSID probe, will therefore not be discovered until the device is recreated.

Would it make sense to merge the current NVRAM VAP list into the cached candidates on each enumeration? This does not appear to block the reported initial-discovery fix.

**Review by wborn — COMMENTED** [review]
Thanks for adding virtual access point discovery. The UID sanitization looks correct, while preserving the original interface name in `interfaceId` keeps radio lookups working. Reading the Broadcom `wl*_vifs` values and validating the returned interface names also looks sensible.

I only found one non-blocking lifecycle concern: after the first successful enumeration, the VAP configuration is no longer checked. This does not block the reported initial-discovery fix, but it may prevent VAPs added or temporarily missed later from being discovered without recreating the device.

This review was AI-assisted.

**Review by wborn — COMMENTED** [review]
Thanks for addressing the previous VAP lifecycle concern. Re-reading the NVRAM VAP list on every enumeration and adding the two-pass test fixes that case.

I found one remaining issue with updating `discoveredIfaces`: a temporary failure while querying one physical radio can now permanently remove that radio from future probes. The current CI run passes, so apart from this I did not find additional issues.

This review was AI-assisted.

File: bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java [hunk 2]
```diff
@@ -141,8 +155,8 @@ protected List<DDWRTRadio> enumerateRadios(SshRunner runner) {
                 logger.debug("Found Broadcom radio: {}", radio);
             }
         }
-        // Cache discovered interfaces to avoid probing non-existent ones on subsequent refreshes
-        if (!foundIfaces.isEmpty() && discoveredIfaces == null) {
+        // Keep physical probes cached while allowing the current NVRAM VAP list to change.
+        if (!foundIfaces.isEmpty()) {
```
Comment (wborn) [inline bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java]: Could we avoid replacing the cached physical interfaces with only the interfaces that succeeded during the latest scan?

For example, after `eth1` and `eth2` have been discovered, if the `eth2` SSID query temporarily returns empty while `eth1` succeeds, this updates `discoveredIfaces` to only `eth1`. Since `BROADCOM_IFACES` is only added when the cache is `null`, `eth2` is never probed again.

It might be better to keep the discovered physical interfaces separately and combine those with the current NVRAM VAP list on each enumeration.

Comment (ballle98) [reply bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java]: Good catch. I now keep the initially discovered physical interfaces in a separate cache and never replace that cache with the latest successful scan. Each enumeration combines those physical interfaces with the current NVRAM VAP list, so a transient `eth2` SSID failure is retried on the next scan. The updated three-pass test covers initial discovery, a temporary `eth2` failure while VAPs are added, and recovery. `mvn clean install` passes with all 66 tests.

**Review by wborn — COMMENTED** [review]
Thanks for addressing the previous physical-interface caching concern. Keeping the discovered physical interfaces separate from the current NVRAM VAP list fixes the later transient-failure case, and the updated three-pass test covers that behavior well.

AI review found one remaining edge case around partial success during the initial physical-interface enumeration: once any physical interface is cached, another physical interface that temporarily failed during that first scan is

**Review by wborn — COMMENTED** [review]
Thanks for addressing the previous physical-interface caching concern. Keeping the discovered physical interfaces separate from the current NVRAM VAP list fixes the later transient-failure case, and the updated three-pass test covers that behavior well.

AI review found one remaining edge case around partial success during the initial physical-interface enumeration: once any physical interface is cached, another physical interface that temporarily failed during that first scan is no longer retried. An inline comment has been added for that case.

Apart from this, no additional issues were found in the current changes.

File: bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java [hunk 3]
```diff
@@ -141,10 +155,13 @@ protected List<DDWRTRadio> enumerateRadios(SshRunner runner) {
...
         }
-        // Cache discovered interfaces to avoid probing non-existent ones on subsequent refreshes
-        if (!foundIfaces.isEmpty() && discoveredIfaces == null) {
-            discoveredIfaces = foundIfaces.toArray(new String[0]);
-            logger.debug("Cached Broadcom interfaces: {}", foundIfaces);
+        // Keep known physical interfaces across transient failures while reading VAPs from NVRAM each time.
+        if (discoveredPhysicalIfaces == null) {
+            List<String> foundPhysicalIfaces = foundIfaces.stream().filter(BROADCOM_IFACES::contains).toList();
+            if (!foundPhysicalIfaces.isEmpty()) {
+                discoveredPhysicalIfaces = foundPhysicalIfaces.toArray(new String[0]);
```
Comment (wborn) [inline bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java]: Could we also keep retrying physical candidates that did not succeed during the initial enumeration?

For example, if `eth1` succeeds but `eth2` temporarily returns an empty SSID on the first scan, `discoveredPhysicalIfaces` is initialized to only `eth1`. From then on `BROADCOM_IFACES` is no longer added to the candidates, so `eth2` is never probed again.

The new test covers a transient `eth2` failure after both physical interfaces have already been cached, but not during the initial cache population. Keeping unresolved physical candidates in the probe set, or otherwise retrying them, would avoid permanently dropping a radio because of a first-scan failure.

Comment (ballle98) [reply bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java]: Good point. I extended the existing combined NVRAM command to read `wl*_ifname` together with `wl*_vifs` on every enumeration, so a physical interface missed during the initial SSID probe is added to the next candidate set without another SSH execution. Successfully discovered physical interfaces are accumulated in the stable cache and are never removed by a transient failure. The updated four-pass test covers an initial `eth2` failure and recovery, then a later cached failure and recovery while VAPs remain refreshed. `mvn clean install` passes with all 66 tests.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

I guess you like this PR being backported ?

File: bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java [hunk 4]
```diff
@@ -113,11 +119,20 @@ protected void refreshWirelessClients(SshRunner runner) {
...
-        String[] ifaces = discoveredIfaces;
+        Set<String> candidates = new LinkedHashSet<>();
+        String[] ifaces = discoveredPhysicalIfaces;
         if (ifaces == null) {
-            ifaces = BROADCOM_IFACES;
+            candidates.addAll(BROADCOM_IFACES);
+        } else {
+            candidates.addAll(List.of(ifaces));
+        }
+        String configuredIfaces = safeTrim(runner.execStdout(GET_CONFIGURED_IFACES_COMMAND));
```
Comment (wborn) [inline bundles/org.openhab.binding.ddwrt/src/main/java/org/openhab/binding/ddwrt/internal/api/DDWRTBroadcomDevice.java:129]: Could we also reconcile VAPs that disappear from the configured interface list?

For example, after `wl0.1` has been discovered, removing that VAP means it is no longer added to `candidates`, but the previously discovered radio remains in `DDWRTNetworkCache` because `refreshRadios()` only adds or updates radios. `refreshWirelessClients()` subsequently iterates the cached radios, so the removed VAP's old assoclist can continue marking clients online, and discovery can continue exposing the stale radio.

It would be good to remove VAP radios that are no longer configured while retaining the current protection against removing physical radios because of transient SSID-query failures.

**Review by wborn — COMMENTED** [review]
Thanks for addressing the previous physical-interface retry concern. Reading the configured `wl*_ifname` values on every enumeration allows physical interfaces missed during the initial SSID probe to be retried, while the stable physical-interface cache continues to protect against transient failures.

AI review found one remaining VAP lifecycle issue around interfaces that are removed after discovery. An inline comment has been added for that case.

Apart from this, no additional issues were found in the final changes.
